plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+}
+ECL_REDIS_API void ECL_REDIS_CALL AsyncRGetUChar(ICodeContext * ctx, size32_t & returnLength, UChar * & returnValue,  const char * options, const char * key, unsigned __int64 database)
+{
+    size_t returnSize = 0;
+    AsyncRGet(ctx, options, key, returnSize, returnValue, NULL, database);
+    returnLength = static_cast<size32_t>(returnSize/sizeof(UChar));
+}
+ECL_REDIS_API void ECL_REDIS_CALL AsyncRGetUtf8(ICodeContext * ctx, size32_t & returnLength, char * & returnValue, const char * options, const char * key, unsigned __int64 database)
+{
+    size_t returnSize = 0;
+    AsyncRGet(ctx, options, key, returnSize, returnValue, NULL, database);
+    returnLength = static_cast<size32_t>(rtlUtf8Length(returnSize, returnValue));
+}
+ECL_REDIS_API void ECL_REDIS_CALL AsyncRGetData(ICodeContext * ctx, size32_t & returnLength, void * & returnValue, const char * options, const char * key, unsigned __int64 database)
+{
+    size_t _returnLength = 0;
+    AsyncRGet(ctx, options, key, _returnLength, returnValue, NULL, database);
+    returnLength = static_cast<size32_t>(_returnLength);
+}
+//----------------------------------------LOCKING WRAPPERS------------------------------------------------------------------------------------------
+ECL_REDIS_API unsigned __int64 ECL_REDIS_CALL RGetLockObject(ICodeContext * ctx, const char * options, const char * key, unsigned __int64 database)
+{
+    Owned<KeyLock> lockObject;
+    StringBuffer channel;
+    channel.set(REDIS_LOCK_PREFIX);
+    channel.append("_").append(key).append("_");
+    channel.append("_").append(database).append("_");
+
+    //srand(unsigned int seed);
+    channel.append(rand()%REDIS_MAX_LOCKS+1);
+
+    lockObject.set(new KeyLock(ctx, options, key, channel.str(), database));
+    return reinterpret_cast<unsigned long long>(lockObject.get());
+}
+ECL_REDIS_API bool ECL_REDIS_CALL RMissThenLock(ICodeContext * ctx, unsigned __int64 _lockObject)
+{
+    KeyLock * lockObject = (KeyLock*)_lockObject;
+    const char * channel = lockObject->getChannel();
+    if (!lockObject || strlen(channel) == 0)
+    {
+        VStringBuffer msg("Redis Plugin : ERROR 'Locking.ExistLockSub' called without sufficient LockObject.");
+        rtlFail(0, msg.str());
+    }
+    const char * options = lockObject->getOptions();
+    Owned<AsyncConnection> master = AsyncConnection::createConnection(ctx, options, lockObject->getDatabase());
+    return master->missThenLock(ctx, lockObject);
+}
+//-----------------------------------SET------------------------------------------
+ECL_REDIS_API void ECL_REDIS_CALL LockingRSetStr(ICodeContext * ctx, size32_t & returnLength, char * & returnValue, unsigned __int64 _lockObject, size32_t valueLength, const char * value, unsigned expire)
+{
+    KeyLock * lockObject = (KeyLock*)_lockObject;
+    AsyncRSet(ctx, lockObject->getOptions(), lockObject->getKey(), valueLength, value, expire, lockObject->getChannel(), lockObject->getDatabase());
```

Comment:
It may not be worth it since it is so small, but the number of lockObject->... suggests this should be a member of KeyLock - if nothing else it is breaking encapsulation.